### ISSUE TITLE
Context manager for startup and shutdown of connection

### ISSUE THREAD

**Issue by caniko** [Intreecom/scyllapy#37]
Looking for this:
```python
with Scylla(...) as scylla:
    <the transactions>
```

`__aenter__` should do `startup` and `__aexist__` should do `shutdown`.

If this is OK, I'd like to implement this.

**Comment by s3rius** [comment]
Actually, you might add this. It shouldn't be problematic to implement.

**Comment by caniko** [comment]
Cool!

I will use it with dependency injection in FastAPI; the connection will shut down naturally when we stop the server or if the WSGI dies for some reason.

**Comment by s3rius** [comment]
Can you provide me with example of how you want to use it? Minimal application.

**Comment by caniko** [comment]
This should work

```python

app = FastAPI()

async def get_scylla_session()  -> AsyncGenerator[Scylla, None]:
    with Scylla(...) as scylla:
        yield scylla

@app.get("/")
def root_endpoint(scylla: Scylla = Depends(get_scylla_session)):
    # scylla can be used to interact with DB
    ...

**Comment by s3rius** [comment]
You should not do so under any circumstances. It is going to be disastrously slow, because it is going to create a connection pool on every request. When you call shutdown it closes all connections to scylla. You don't want such thing.

So, that's how you actually need to use it:


```python
from fastapi import Depends, FastAPI, Request
from typing import Any, AsyncGenerator
from contextlib import asynccontextmanager

from scyllapy import Scylla


@asynccontextmanager
async def lifespan(app: FastAPI) -> AsyncGenerator[None, None]:
    print("Starting up")
    scylla = Scylla(
        ["localhost:9042"],
        username="scylla",
        password="scylla",
        keyspace="test",
    )
    await scylla.startup()
    app.state.scylla = scylla

    yield

    await scylla.shutdown()

    print("Shutting down")


app = FastAPI(lifespan=lifespan)


async def get_scylla_session(request: Request) -> Scylla:
    return request.app.state.scylla


@app.get("/")
async def root_endpoint(
    scylla: Scylla = Depends(get_scylla_session),
) -> list[dict[str, Any]]:
    res = await scylla.execute("SELECT * FROM test")
    return res.all()
```

Here we initialize pool only once and then reuse connections. We store the initialized scylla in the state, so we can access it from dependencies and use in our handlers. This setup is going to work many times faster.

**Comment by caniko** [comment]
Thanks for the suggestion!

> because it is going to create a connection pool on every request

This is false, I don't think you know how it works, read about it [here](https://fastapi.tiangolo.com/tutorial/dependencies/)

**Comment by caniko** [comment]
Just tested this, you are correct. After this PR, the code should be like this:
```python
from fastapi import Depends, FastAPI, Request
from typing import Any, AsyncGenerator
from contextlib import asynccontextmanager

from scyllapy import Scylla


@asynccontextmanager
async def lifespan(app: FastAPI) -> AsyncGenerator[None, None]:
    print("Starting up")
    scylla = Scylla(
        ["localhost:9042"],
        username="scylla",
        password="scylla",
        keyspace="test",
    )
    app.state.scylla = scylla

    with Scylla(...):
        yield


app = FastAPI(lifespan=lifespan)


async def get_scylla_session(request: Request) -> Scylla:
    return request.app.state.scylla


@app.get("/")
async def root_endpoint(
    scylla: Scylla = Depends(get_scylla_session),
) -> list[dict[str, Any]]:
    res = await scylla.execute("SELECT * FROM test")
    return res.all()
```

**Comment by Arap-Odawa** [comment]
Dear @caniko and @s3rius ,

I hope that you're all keeping well.

First of all, I respect all your collective achievements. From your Github repositories, I can tell that you guys are technically excellent, I hope to be at the levels where you guys are at in 1-2 years.

@caniko , I had sent you an email on 4th February, 2024 to request you to send me the startup, shutdown and setup of the database in terms of the docker compose and the scylla.yaml file and also the db.py equivalent so that I can be able to customize a ScyllaDB connection for a project I am working on. In December, 2023, had also made a request to you @s3rius  to work on a FastAPI boilerplate template that would assist newbies like me in project setup and also may increase the popularity of the scyllapy driver.

What I've been able o do since the email to @caniko: I've been able to complete three Scylla University courses on ScyllaDB and also read through nearly all available material on ScyllaDB. 

What I've done with the knowledge: I've been able to get to the stage where I can setup the startup and shutdown events for the scyllapy connection but it does not connect. I get the following error as at yesternight:

await _client.startup()
scyllapy.exceptions.ScyllaPySessionError: Cannot construct new session: IO Error: 
No connections in the pool; last connection failed with: Invalid message: Frame error: early eof.

On top of the above error message, I have the following questions to @s3rius  or @caniko :

1. How would you use models to create tables while setting up scyllapy-based application? Is there a need for the sync_table while using scyllapy?
2. Ho do you connect to a multi-node cluster? The example shown above connects to a single node, how can you modify it to connect to a 5 node cluster?
3. What could be wrong with the following db connection statements? Is there a need for the session?

```
from scyllapy import Consistency, ExecutionProfile, Query, Scylla, SerialConsistency
from scyllapy.load_balancing import LoadBalancingPolicy, LatencyAwareness

from utils.config import (get_settings)  

from typing import NewType

from cassandra.cluster import Cluster, Session

DBSession = NewType("DBSession", Session)


_client: DBSession = None

settings = get_settings()

default_profile = ExecutionProfile(
    consistency=Consistency.LOCAL_QUORUM,
    serial_consistency=SerialConsistency.LOCAL_SERIAL,
    request_timeout=2,
)

async def get_client():
    return _client

async def connect_to_database():

    global _client

    _client = Scylla(
#    key_space = settings.SCYLLA_KEYSPACE,
#    contact_points = [
#        f"{settings.SCYLLA_KEYSPACE}.{settings.SCYLLA_DB_NODE_1}:{settings.SCYLLA_DB_NODE_1_PORT}",
#        f"{settings.SCYLLA_KEYSPACE}.{settings.SCYLLA_DB_NODE_2}:{settings.SCYLLA_DB_NODE_2_PORT}",
#        f"{settings.SCYLLA_KEYSPACE}.{settings.SCYLLA_DB_NODE_3}:{settings.SCYLLA_DB_NODE_3_PORT}",
#        f"{settings.SCYLLA_KEYSPACE}.{settings.SCYLLA_DB_NODE_4}:{settings.SCYLLA_DB_NODE_4_PORT}",
#        f"{settings.SCYLLA_KEYSPACE}.{settings.SCYLLA_DB_NODE_5}:{settings.SCYLLA_DB_NODE_5_PORT}",
#                      
#                      ],

    contact_points = [
        f"{settings.SCYLLA_DB_HOST}:{settings.SCYLLA_DB_NODE_1_PORT}",
        f"{settings.SCYLLA_DB_HOST}:{settings.SCYLLA_DB_NODE_2_PORT}",
        f"{settings.SCYLLA_DB_HOST}:{settings.SCYLLA_DB_NODE_3_PORT}",
        f"{settings.SCYLLA_DB_HOST}:{settings.SCYLLA_DB_NODE_4_PORT}",
        f"{settings.SCYLLA_DB_HOST}:{settings.SCYLLA_DB_NODE_5_PORT}",
                      
                      ],

    username = settings.SCYLLA_DB_USERNAME,
    password = settings.SCYLLA_DB_PASSWORD,
    default_execution_profile=default_profile,
    )
    await _client.startup()
    await _client.execute(
        f"CREATE KEYSPACE IF NOT EXISTS {settings.SCYLLA_KEYSPACE} WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 5};"
    )
    await _client.use_keyspace(settings.SCYLLA_KEYSPACE)
    

async def close_connection_to_database():
    _client.shutdown()
```

5. How can the above be connected to the code piece that you have shown above to complete the connection to ScyllaDB so that the connection can be complete? Or alternatively, how can the below amended to complete the connection loop to avoid the errors?

```
from model import CheckTable

@asynccontextmanager
async def lifespan_scylla(app: FastAPI):
    # Setup the ScyllaDB Connection
    await connect_to_database()  # connect to database mentioned in the earlier code block
    conn = await get_client()
    
    sync_table(CheckTable)

    yield
    # Close the ScyllaDB Connection
    close_connection_to_database()

    

app = FastAPI(lifespan=lifespan)
```


Your kind responses to this will solve all the problems that I had requested you to assist me with and also reduce the time I have taken in setup of the connection bit. Otherwise, I respect all your collective achievements, from your Github repositories, I can tell that you guys are technically excellent, I hope to be at the levels where you guys are at in 1-2 years.

Thanks,

Regards,

Otieno

**Comment by Arap-Odawa** [comment]
Alternatively, @caniko and @s3rius  how would:

1. I incorporate creating tables using models in the script that you have shown below (is there a role for sync tables, or they are moot when using scyllapy?)
2. Incorporate creation of keyspaces as I have described, i.e. if not exists
3. Incorporate a multi-node contact points on the code that you've shown below


`from fastapi import Depends, FastAPI, Request
from typing import Any, AsyncGenerator
from contextlib import asynccontextmanager

from scyllapy import Scylla


@asynccontextmanager
async def lifespan(app: FastAPI) -> AsyncGenerator[None, None]:
    print("Starting up")
    scylla = Scylla(
        ["localhost:9042"],
        username="scylla",
        password="scylla",
        keyspace="test",
    )
    app.state.scylla = scylla

    with scylla:
        yield


app = FastAPI(lifespan=lifespan)


async def get_scylla_session(request: Request) -> Scylla:
    return request.app.state.scylla


@app.get("/")
async def root_endpoint(
    scylla: Scylla = Depends(get_scylla_session),
) -> list[dict[str, Any]]:
    res = await scylla.execute("SELECT * FROM test")
    return res.all()`

**Comment by Arap-Odawa** [comment]
I've tried requesting ChatGPT (though I don't like using it, feels like cheating) to combine the following two code sections that would kind of sort out the problem that I'm facing but it seems that it is not combining the two into a coherent, workable code piece that can be used to setup a code piece:

code1:

```
from fastapi import Depends, FastAPI, Request
from typing import Any, AsyncGenerator
from contextlib import asynccontextmanager

from scyllapy import Scylla


@asynccontextmanager
async def lifespan(app: FastAPI) -> AsyncGenerator[None, None]:
    print("Starting up")
    scylla = Scylla(
        ["localhost:9042"],
        username="scylla",
        password="scylla",
        keyspace="test",
    )
    app.state.scylla = scylla

    with scylla:
        yield


app = FastAPI(lifespan=lifespan)


async def get_scylla_session(request: Request) -> Scylla:
    return request.app.state.scylla


@app.get("/")
async def root_endpoint(
    scylla: Scylla = Depends(get_scylla_session),
) -> list[dict[str, Any]]:
    res = await scylla.execute("SELECT * FROM test")
    return res.all()
```

and code 2, the table model:

```
from uuid import UUID

from cassandra.cqlengine import columns
from cassandra.cqlengine.models import Model


class Task(Model):
    __table_name__ = "tasks"
    __keyspace__ = "todo_keyspace"

    id = columns.UUID(primary_key=True)
    title = columns.Text(min_length=3, max_length=255)
    description = columns.Text(min_length=0, max_length=512)
    is_completed = columns.Boolean(default=False)
```

**Comment by Arap-Odawa** [comment]
After further review of the scyllapy documentation, it's now clearer in terms of how I can be able to work on the keyspaces. The only bit that is not clear is how I can be able to create tables using models i.e. how to implement a models based table definition for the below snippet of code obtained from the scyllapy documentation:


```
from scyllapy import Scylla
from scyllapy.query_builder import Insert, Select, Update, Delete


async def main(scylla: Scylla):
    await scylla.execute("CREATE TABLE users(id INT PRIMARY KEY, name TEXT)")

```

**Comment by caniko** [comment]
I got your emails, was going to answer; however, this is too much. Can you condense your question, and remove it from here?

Can we turn on the discussion section for future cases @s3rius ?

**Comment by Arap-Odawa** [comment]
> I got your emails, was going to answer; however, this is too much. Can you condense your question, and remove it from here?
> 
> Can we turn on the discussion section for future cases @s3rius ?

Thank you very much @caniko  for taking some time to look into this. 

Just to condense the questions (2 of them):

1. **Question1:** How do I utilise a models based method to create tables while using scyllapy, ie convert the following (snip1) into a model-based statement that can create tables as detailed in snip 2:

Snip 1:

```
from scyllapy import Scylla
from scyllapy.query_builder import Insert, Select, Update, Delete


async def main(scylla: Scylla):
    await scylla.execute("CREATE TABLE users(id INT PRIMARY KEY, name TEXT)")
```

and snip 2:

```
from uuid import UUID

from cassandra.cqlengine import columns
from cassandra.cqlengine.models import Model

class Task(Model):
    __table_name__ = "tasks"
    __keyspace__ = "todo_keyspace"

    id = columns.UUID(primary_key=True)
    title = columns.Text(min_length=3, max_length=255)
    description = columns.Text(min_length=0, max_length=512)
    is_completed = columns.Boolean(default=False)
```

**Question 2: I'm getting the following error in docker for all of my scylla nodes:**

```
2024-02-21 07:20:22 running: (['/opt/scylladb/scripts/scylla_dev_mode_setup', '--developer-mode', '1'],)
2024-02-21 07:20:23 running: (['/opt/scylladb/scripts/scylla_io_setup'],)
2024-02-21 07:20:23 2024-02-21 04:20:23,683 CRIT Supervisor is running as root.  Privileges were not dropped because no user is specified in the config file.  If you intend to run as root, you can set user=root in the config file to avoid this message.
2024-02-21 07:20:23 2024-02-21 04:20:23,684 INFO Included extra file "/etc/supervisord.conf.d/rsyslog.conf" during parsing
2024-02-21 07:20:23 2024-02-21 04:20:23,684 INFO Included extra file "/etc/supervisord.conf.d/scylla-housekeeping.conf" during parsing
2024-02-21 07:20:23 2024-02-21 04:20:23,684 INFO Included extra file "/etc/supervisord.conf.d/scylla-jmx.conf" during parsing
2024-02-21 07:20:23 2024-02-21 04:20:23,684 INFO Included extra file "/etc/supervisord.conf.d/scylla-node-exporter.conf" during parsing
2024-02-21 07:20:23 2024-02-21 04:20:23,684 INFO Included extra file "/etc/supervisord.conf.d/scylla-server.conf" during parsing
2024-02-21 07:20:23 2024-02-21 04:20:23,684 INFO Included extra file "/etc/supervisord.conf.d/sshd-server.conf" during parsing
2024-02-21 07:20:23 2024-02-21 04:20:23,689 INFO RPC interface 'supervisor' initialized
2024-02-21 07:20:23 2024-02-21 04:20:23,689 CRIT Server 'inet_http_server' running without any HTTP authentication checking
2024-02-21 07:20:24 rsyslogd: imklog: cannot open kernel log (/proc/kmsg): Operation not permitted.
2024-02-21 07:20:24 rsyslogd: activation of module imklog failed [v8.2001.0 try https://www.rsyslog.com/e/2145 ]
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.727Z caller=node_exporter.go:182 level=info msg="Starting node_exporter" version="(version=1.3.1, branch=HEAD, revision=a2321e7b940ddcff26873612bccdf7cd4c42b6b6)"
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.727Z caller=node_exporter.go:183 level=info msg="Build context" build_context="(go=go1.17.3, user=root@243aafa5525c, date=20211205-11:09:49)"
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.727Z caller=node_exporter.go:185 level=warn msg="Node Exporter is running as root user. This exporter is designed to run as unpriviledged user, root is not required."
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.728Z caller=filesystem_common.go:111 level=info collector=filesystem msg="Parsed flag --collector.filesystem.mount-points-exclude" flag=^/(dev|proc|run/credentials/.+|sys|var/lib/docker/.+)($|/)
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.728Z caller=filesystem_common.go:113 level=info collector=filesystem msg="Parsed flag --collector.filesystem.fs-types-exclude" flag=^(autofs|binfmt_misc|bpf|cgroup2?|configfs|debugfs|devpts|devtmpfs|fusectl|hugetlbfs|iso9660|mqueue|nsfs|overlay|proc|procfs|pstore|rpc_pipefs|securityfs|selinuxfs|squashfs|sysfs|tracefs)$
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:108 level=info msg="Enabled collectors"
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=arp
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=bcache
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=bonding
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=btrfs
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=conntrack
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=cpu
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=cpufreq
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=diskstats
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=dmi
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=edac
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=entropy
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=fibrechannel
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=filefd
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=filesystem
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=hwmon
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=infiniband
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=interrupts
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=ipvs
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=loadavg
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=mdadm
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=meminfo
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=netclass
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=netdev
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=netstat
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=nfs
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=nfsd
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=nvme
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=os
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=powersupplyclass
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=pressure
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=rapl
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=schedstat
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=sockstat
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=softnet
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=stat
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=tapestats
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=textfile
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=thermal_zone
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=time
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=timex
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=udp_queues
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=uname
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=vmstat
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=xfs
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.730Z caller=node_exporter.go:115 level=info collector=zfs
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.731Z caller=node_exporter.go:199 level=info msg="Listening on" address=:9100
2024-02-21 07:20:24 ts=2024-02-21T04:20:24.731Z caller=tls_config.go:195 level=info msg="TLS is disabled." http2=false
2024-02-21 07:20:24 FATAL: Exception during startup, aborting: std::runtime_error (Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application)
2024-02-21 07:20:23 2024-02-21 04:20:23,690 INFO supervisord started with pid 26
2024-02-21 07:20:24 2024-02-21 04:20:24,698 INFO spawned: 'rsyslog' with pid 28
2024-02-21 07:20:24 2024-02-21 04:20:24,702 INFO spawned: 'scylla' with pid 29
2024-02-21 07:20:24 2024-02-21 04:20:24,707 INFO spawned: 'scylla-housekeeping' with pid 32
2024-02-21 07:20:24 2024-02-21 04:20:24,710 INFO spawned: 'scylla-jmx' with pid 33
2024-02-21 07:20:24 2024-02-21 04:20:24,712 INFO spawned: 'scylla-node-exporter' with pid 35
2024-02-21 07:20:24 2024-02-21 04:20:24,713 INFO spawned: 'sshd' with pid 37
2024-02-21 07:20:24 2024-02-21 04:20:24,945 INFO exited: scylla (exit status 7; not expected)
2024-02-21 07:20:25 Connecting to http://localhost:10000
2024-02-21 07:20:25 Starting the JMX server
2024-02-21 07:20:25 JMX is enabled to receive remote connections on port: 7199
2024-02-21 07:20:25 2024-02-21 04:20:25,716 INFO success: rsyslog entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2024-02-21 07:20:25 2024-02-21 04:20:25,716 INFO success: scylla-housekeeping entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2024-02-21 07:20:25 2024-02-21 04:20:25,716 INFO success: scylla-jmx entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2024-02-21 07:20:25 2024-02-21 04:20:25,716 INFO success: scylla-node-exporter entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2024-02-21 07:20:25 2024-02-21 04:20:25,716 INFO success: sshd entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2024-02-21 07:20:26 2024-02-21 04:20:26,718 INFO spawned: 'scylla' with pid 76
2024-02-21 07:20:26 2024-02-21 04:20:26,841 INFO exited: scylla (exit status 7; not expected)
2024-02-21 07:20:28 2024-02-21 04:20:28,844 INFO spawned: 'scylla' with pid 87
2024-02-21 07:20:28 2024-02-21 04:20:28,942 INFO exited: scylla (exit status 7; not expected)
2024-02-21 07:20:32 2024-02-21 04:20:32,098 INFO spawned: 'scylla' with pid 110
2024-02-21 07:20:32 2024-02-21 04:20:32,203 INFO exited: scylla (exit status 7; not expected)
2024-02-21 07:20:33 2024-02-21 04:20:33,204 INFO gave up: scylla entered FATAL state, too many start retries too quickly
2024-02-21 07:20:26 FATAL: Exception during startup, aborting: std::runtime_error (Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application)
2024-02-21 07:20:28 FATAL: Exception during startup, aborting: std::runtime_error (Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application)
2024-02-21 07:20:30 Traceback (most recent call last):
2024-02-21 07:20:30   File "/opt/scylladb/scripts/libexec/scylla-housekeeping", line 196, in <module>
2024-02-21 07:20:30     args.func(args)
2024-02-21 07:20:30   File "/opt/scylladb/scripts/libexec/scylla-housekeeping", line 122, in check_version
2024-02-21 07:20:30     current_version = sanitize_version(get_api('/storage_service/scylla_release_version'))
2024-02-21 07:20:30   File "/opt/scylladb/scripts/libexec/scylla-housekeeping", line 80, in get_api
2024-02-21 07:20:30     return get_json_from_url("http://" + api_address + path)
2024-02-21 07:20:30   File "/opt/scylladb/scripts/libexec/scylla-housekeeping", line 75, in get_json_from_url
2024-02-21 07:20:30     raise RuntimeError(f'Failed to get "{path}" due to the following error: {retval}')
2024-02-21 07:20:30 RuntimeError: Failed to get "http://localhost:10000/storage_service/scylla_release_version" due to the following error: <urlopen error [Errno 99] Cannot assign requested address>
2024-02-21 07:20:32 FATAL: Exception during startup, aborting: std::runtime_error (Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application)
```
I've tried setting the aio-max-nr to the 1045678 but it's not working. Any leads on how I can be able to handle this? My setup is running docker on Windows.

Many thanks and thanks for your time,

Regards,

Odawa

**Comment by Arap-Odawa** [comment]
Hi @caniko ,

I have migrated my question to a new issue: [https://github.com/Intreecom/scyllapy/issues/42](url)

Thanks and apologies for the earlier many postings, I have cleaned them up.

Regards,

Odawa
